### PR TITLE
emit cargo metadata duiring build scripts to avoid outdated buildscript outputs

### PR DESCRIPTION
**Pull request by pascalkuthe**
While triaging issues I often noticed that the revision reported by `helix --version` if often wrong. I finally found the reason for this: The `build.rs` script that extracts the HEAD branch by calling `git` is not automatically rerun so if you build helix then checkout/merge a branch/revision that doesn't touch `helix-loader` (which doesn't see that much activity compared to the rest of the editor) and rebuild then build script will not be rerun and`helix --version` will report the old (incorrect) revision.

I made the build script emit an appropriate `cargo:rerun-if-changed=` so that it is rerun if `HEAD` changes to fix that.

### CODE REVIEW

**Review by the-mikedavis — DISMISSED** [review]

**Review by dead10ck — CHANGES_REQUESTED** [review]
Oh shoot, I had stumbled across this myself a long time ago and then totally forgot to make a PR! I think we also have to rerun if basically any file changes inside `runtime/grammars/sources`, or the `languages.toml` file too.

**Comment by pascalkuthe** [comment]
> 

I looked into adding this. We are already rebuilding on `languages.toml` changes since we `include_bytes!` those which is a compiler builtin that handles invalidation correctly.

I added the metadata for `runtime/grammars/sources`. In the process I noticed that the simple approach I had taken to git metadata before didn't quite work and I didn't consider that HEAD is (usually) a git symbolic-ref and that `.git` can also be a reference (for detached work trees which I am even using myself). Considering that I heavily contribute to gitoxide I really should know better but it slipped me by. The new solution is much more robust. Sorry about that

File: helix-loader/build.rs [hunk 1]
```diff
@@ -23,4 +24,40 @@ fn main() {
...
+    // from command output so for now we can't accept non-utf8 paths here
+    // probably rare enouch where it doesn't matter tough we could use gitoxide
+    // here but that would be make it a hard dependency and slow compile times
+    let Some(git_dir): Option<String> = Command::new("git")
+        .args(["rev-parse", "--git-dir"])
+        .output()
+        .ok()
+        .filter(|output| output.status.success())
+        .and_then(|x| String::from_utf8(x.stdout).ok())
+    else{ return; };
```
Comment (dead10ck) [inline helix-loader/build.rs:43]: Does `cargo fmt` work on build scripts?

**Review by dead10ck — APPROVED** [review]
Ahh, that makes sense, thanks! This looks good to me.

File: helix-loader/build.rs (hunk 1 above)
Comment (pascalkuthe) [reply helix-loader/build.rs:43]: yes but rustfmt currently ignores `let .. else` (leaves it unchanged) altough that will hopefully be fixed in the not-so-distant future

**Review by the-mikedavis — APPROVED** [review]

**Pull request merged**